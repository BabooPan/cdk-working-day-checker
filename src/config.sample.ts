export interface recordProps {
  /**
     * The record goona be created in the specific domain.
     */
  readonly recordName: string;
  /**
     * The domain name goona be used to create record and where the cert in.
     */
  readonly domainName: string;
  /**
     * The domain id from Route53 goona be used to create record and where the cert in.
     */
  readonly zoneId: string;
  /**
     * The ACM ARN specific to the API GW custom domain name
     */
  readonly certArn: string;
}

export const record:recordProps = {
  recordName: 'record',
  domainName: 'example.com',
  zoneId: 'Zxxxxxxx01234',
  certArn: 'arn:aws:acm:REGION_CODE:ACCOUNT_ID:certificate/RESOURCE_ID',
};
