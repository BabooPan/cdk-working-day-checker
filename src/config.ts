export interface RecordProps {
  /**
     * The record goona be created in the specific domain.
     */
  readonly recordName: string;
  /**
     * The domain name goona be used to create record and where the cert in.
     */
  readonly domainName: string;
  /**
     * The domain id from Route53 goona be used to create record and where the cert in.
     */
  readonly zoneId: string;
  /**
     * The ACM ARN specific to the API GW custom domain name
     */
  readonly certArn: string;
}

export const record:RecordProps = {
  recordName: 'workingday',
  domainName: 'baboopan.net',
  zoneId: 'Z00292561235P7FTLUPZ9',
  certArn: 'arn:aws:acm:us-west-2:471856162574:certificate/63d81b75-628d-4ee7-8696-46f619e8c77e',
};
