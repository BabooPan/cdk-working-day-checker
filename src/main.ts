import { record } from './config'
import { App, Construct, Stack, StackProps, CfnOutput } from '@aws-cdk/core';
import { PythonFunction } from '@aws-cdk/aws-lambda-python';
import { Certificate } from '@aws-cdk/aws-certificatemanager';
import * as lambda from '@aws-cdk/aws-lambda';
import * as apigwv2 from '@aws-cdk/aws-apigatewayv2';
import * as apigwInt from '@aws-cdk/aws-apigatewayv2-integrations';
import * as route53 from '@aws-cdk/aws-route53';
import * as targets from '@aws-cdk/aws-route53-targets';
// import * as path from 'path';

export class CdkWorkingDayChecker extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    /** 
     * Lack support for custom modules in Python
     */
    // const WorkingDayCheckerFunction = new lambda.Function(this, 'workingDayCheckerFunction', {
    //   runtime: lambda.Runtime.PYTHON_3_9,
    //   code: lambda.Code.fromAsset(path.join('./lambda-handler')),
    //   handler: 'index.lambda_handler',
    //   architectures: [lambda.Architecture.ARM_64]
    // });
    
    const WorkingDayCheckerFunction = new PythonFunction(this, 'workingDayCheckerFunction', {
      entry: './lambda-handler/',
      index: 'index.py',
      handler: 'lambda_handler',
      runtime: lambda.Runtime.PYTHON_3_9,
      // Due to local device is base on x86, un-support local test for arm64
      // architectures: [lambda.Architecture.ARM_64]
    });

    // const getCheckerIntegration = new apigwInt.LambdaProxyIntegration({
    const getCheckerIntegration = new apigwInt.HttpLambdaIntegration(
      'checkerIntegration',
      WorkingDayCheckerFunction
    );
    
    const apiCustomDomain = new apigwv2.DomainName(this, 'apiCustomDomain', {
      domainName: record.recordName + '.' + record.domainName,
      certificate: Certificate.fromCertificateArn(this, 'cert', record.certArn),
    });
    
    const httpApi = new apigwv2.HttpApi(this, 'workingDayCheckerApi', {
      createDefaultStage: false
    });

    httpApi.addRoutes({
      path: '/',
      methods: [ apigwv2.HttpMethod.GET ],
      integration: getCheckerIntegration
    });

    const apiStage = httpApi.addStage('dev', {
      stageName: 'dev',
      autoDeploy: true,
      domainMapping: {
        domainName: apiCustomDomain
      },
    });

    const hostedZone = route53.PublicHostedZone.fromHostedZoneAttributes(this, 'HostedZone',{ 
      zoneName: record.domainName,
      hostedZoneId: record.zoneId
    });
    
    new route53.ARecord(this, 'AliasRecord', {
      zone: hostedZone,
      recordName: record.recordName,
      target: route53.RecordTarget.fromAlias(new targets.ApiGatewayv2DomainProperties(apiCustomDomain.regionalDomainName, apiCustomDomain.regionalHostedZoneId))
    });

    const stageCfnResource = apiStage.node.defaultChild as apigwv2.CfnStage;
    stageCfnResource.addPropertyOverride('DefaultRouteSettings', { ThrottlingRateLimit: 5 });

    new CfnOutput(this, 'apigw-endpoint',{ value: httpApi.apiEndpoint + '/dev/'});
    new CfnOutput(this, 'openapi-endpoint',{ value: ('https://' + record.recordName + '.' + record.domainName)});
  }
}

// for development, use account/region from cdk cli
const devEnv = {
  account: process.env.CDK_DEFAULT_ACCOUNT,
  region: process.env.CDK_DEFAULT_REGION,
};

const app = new App();

new CdkWorkingDayChecker(app, 'cdk-working-day-checker', { env: devEnv });

app.synth();