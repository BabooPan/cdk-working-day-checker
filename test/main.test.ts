import '@aws-cdk/assert/jest';
import { App } from '@aws-cdk/core';
import { Template } from '@aws-cdk/assertions';
import { CdkWorkingDayChecker } from '../src/main';

test('Snapshot', () => {
  const app = new App();
  const stack = new CdkWorkingDayChecker(app, 'test');

  const template = Template.fromStack(stack);
  expect(template.toJSON()).toMatchSnapshot();
});