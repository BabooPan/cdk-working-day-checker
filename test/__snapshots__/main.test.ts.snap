// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
Object {
  "Outputs": Object {
    "apigwendpoint": Object {
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            Object {
              "Fn::GetAtt": Array [
                "workingDayCheckerApiE19AE4A3",
                "ApiEndpoint",
              ],
            },
            "/dev/",
          ],
        ],
      },
    },
    "openapiendpoint": Object {
      "Value": "https://workingday.baboopan.net",
    },
  },
  "Parameters": Object {
    "AssetParameters0199382c47039b5f03b76109a943a435cc497fe8d8ea9793b38acfee5b8697e6ArtifactHash67CE28B7": Object {
      "Description": "Artifact hash for asset \\"0199382c47039b5f03b76109a943a435cc497fe8d8ea9793b38acfee5b8697e6\\"",
      "Type": "String",
    },
    "AssetParameters0199382c47039b5f03b76109a943a435cc497fe8d8ea9793b38acfee5b8697e6S3Bucket34D5D11B": Object {
      "Description": "S3 bucket for asset \\"0199382c47039b5f03b76109a943a435cc497fe8d8ea9793b38acfee5b8697e6\\"",
      "Type": "String",
    },
    "AssetParameters0199382c47039b5f03b76109a943a435cc497fe8d8ea9793b38acfee5b8697e6S3VersionKey2917BAEC": Object {
      "Description": "S3 key for asset version \\"0199382c47039b5f03b76109a943a435cc497fe8d8ea9793b38acfee5b8697e6\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "AliasRecord851000D2": Object {
      "Properties": Object {
        "AliasTarget": Object {
          "DNSName": Object {
            "Fn::GetAtt": Array [
              "apiCustomDomain9799E20D",
              "RegionalDomainName",
            ],
          },
          "HostedZoneId": Object {
            "Fn::GetAtt": Array [
              "apiCustomDomain9799E20D",
              "RegionalHostedZoneId",
            ],
          },
        },
        "HostedZoneId": "Z00292561235P7FTLUPZ9",
        "Name": "workingday.baboopan.net.",
        "Type": "A",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "apiCustomDomain9799E20D": Object {
      "Properties": Object {
        "DomainName": "workingday.baboopan.net",
        "DomainNameConfigurations": Array [
          Object {
            "CertificateArn": "arn:aws:acm:us-west-2:471856162574:certificate/63d81b75-628d-4ee7-8696-46f619e8c77e",
            "EndpointType": "REGIONAL",
          },
        ],
      },
      "Type": "AWS::ApiGatewayV2::DomainName",
    },
    "workingDayCheckerApiE19AE4A3": Object {
      "Properties": Object {
        "Name": "workingDayCheckerApi",
        "ProtocolType": "HTTP",
      },
      "Type": "AWS::ApiGatewayV2::Api",
    },
    "workingDayCheckerApiGET37B44C5E": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "workingDayCheckerApiE19AE4A3",
        },
        "AuthorizationType": "NONE",
        "RouteKey": "GET /",
        "Target": Object {
          "Fn::Join": Array [
            "",
            Array [
              "integrations/",
              Object {
                "Ref": "workingDayCheckerApiGETcheckerIntegration62A9726F",
              },
            ],
          ],
        },
      },
      "Type": "AWS::ApiGatewayV2::Route",
    },
    "workingDayCheckerApiGETcheckerIntegration62A9726F": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "workingDayCheckerApiE19AE4A3",
        },
        "IntegrationType": "AWS_PROXY",
        "IntegrationUri": Object {
          "Fn::GetAtt": Array [
            "workingDayCheckerFunction6B548FB9",
            "Arn",
          ],
        },
        "PayloadFormatVersion": "2.0",
      },
      "Type": "AWS::ApiGatewayV2::Integration",
    },
    "workingDayCheckerApiGETcheckerIntegrationPermissionB3DD5A74": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "workingDayCheckerFunction6B548FB9",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "workingDayCheckerApiE19AE4A3",
              },
              "/*/*/",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "workingDayCheckerApidev8CC58CB6": Object {
      "DependsOn": Array [
        "apiCustomDomain9799E20D",
      ],
      "Properties": Object {
        "ApiId": Object {
          "Ref": "workingDayCheckerApiE19AE4A3",
        },
        "AutoDeploy": true,
        "DefaultRouteSettings": Object {
          "ThrottlingRateLimit": 5,
        },
        "StageName": "dev",
      },
      "Type": "AWS::ApiGatewayV2::Stage",
    },
    "workingDayCheckerApidevtestapiCustomDomainundefinedE701E8C0": Object {
      "DependsOn": Array [
        "apiCustomDomain9799E20D",
        "workingDayCheckerApidev8CC58CB6",
      ],
      "Properties": Object {
        "ApiId": Object {
          "Ref": "workingDayCheckerApiE19AE4A3",
        },
        "DomainName": Object {
          "Ref": "apiCustomDomain9799E20D",
        },
        "Stage": "dev",
      },
      "Type": "AWS::ApiGatewayV2::ApiMapping",
    },
    "workingDayCheckerFunction6B548FB9": Object {
      "DependsOn": Array [
        "workingDayCheckerFunctionServiceRoleEDDC65F0",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters0199382c47039b5f03b76109a943a435cc497fe8d8ea9793b38acfee5b8697e6S3Bucket34D5D11B",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters0199382c47039b5f03b76109a943a435cc497fe8d8ea9793b38acfee5b8697e6S3VersionKey2917BAEC",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters0199382c47039b5f03b76109a943a435cc497fe8d8ea9793b38acfee5b8697e6S3VersionKey2917BAEC",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "index.lambda_handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "workingDayCheckerFunctionServiceRoleEDDC65F0",
            "Arn",
          ],
        },
        "Runtime": "python3.9",
      },
      "Type": "AWS::Lambda::Function",
    },
    "workingDayCheckerFunctionServiceRoleEDDC65F0": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
